src/main/scala/com/fulcrumgenomics/math/BinomialDistribution.scala

```diff
@@ -0,0 +1,145 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2017 Fulcrum Genomics LLC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.math
+
+import com.fulcrumgenomics.FgBioDef._
+
+import scala.math.{BigDecimal, BigInt}
+
+/**
+  * A high-precision implementation of the math for binomial probabilities.  This implementation is several
+  * times slower (though not orders of magnitude slower) than the implementation in Apache Commons Math, but
+  * retains precision throughout the computation where the Commons implementation underflows.
+  *
+  * One implementation choice to be aware of when using this implementation is that each instance of the
+  * class will calculate and cache factorials up to factorial(n) where n is the highest value of n
+  * supplied to any call to [[probability()]] or [[cumulativeProbability()]].  Thus to ensure reasonable
+  * performance it is recommended to create one instance of this class and reuse it for as may
+  * calculations as is practical.
+  */
+class BinomialDistribution {
+  // This array will get expanded with more factorials any time a higher n is queried
+  private var factorials = Array(BigInt(0), BigInt(1))
+
+  // Constant values for 0 and 1 in BigDecimal to avoid making them all the time
+  private val Zero = BigDecimal(0)
+  private val One  = BigDecimal(1)
+
+  /** Get the factorial of n. Retrieves the value from a cached set of factorials. If the cache
+    * does not contain factorial(n) yet, then all factorials up to factorial(n) are computed
+    * and cached.
+    *
+    * @param n the number to compute the factorial of
+    * @return the factorial value as a BigInt
+    */
+  private def factorial(n: Int): BigInt = {
+    if (n > factorials.length - 1) {
+      // Expand the array
+      val oldLength = factorials.length
+      val tmp = new Array[BigInt](n+1)
+      System.arraycopy(factorials, 0, tmp, 0, oldLength)
```

Comment:
Is it so much better to use a Array than a List? I was a little horrified by the arraycopy here... :-)

Reply:
_So much better_ might be a stretch, but I don't understand why seeing an arraycopy horrified you?  The array is private within the class, and all it's management is internal to one method.  This is similar to using, e.g. `IndexedSeq`, except that when I need to grow the array it's a one-time operation instead of doubling each time.